File: src/core/speech-queue.js
class SpeechQueue {
  constructor(deliver, options = {}) {
    if (typeof deliver !== 'function') throw new TypeError('SpeechQueue requires a deliver function');
    this.deliver = deliver;
    this.setTimer = options.setTimer || setTimeout;
    this.clearTimer = options.clearTimer || clearTimeout;
    this.pending = [];
    this.current = null;
    this.timer = null;
    this.sequence = 0;
  }

  enqueue(item) {
    if (!item || typeof item.text !== 'string' || item.text.length === 0) return false;
    const entry = {
      ...item,
      priority: Number.isFinite(item.priority) ? item.priority : 10,
      durationMs: Number.isFinite(item.durationMs) ? item.durationMs : 4000,
      sequence: this.sequence++,
    };

    if (entry.replaceKey) {
      this.pending = this.pending.filter((pending) => pending.replaceKey !== entry.replaceKey);
    }

    if (!this.current) {
      this.start(entry);
    } else if (entry.replaceKey && entry.replaceKey === this.current.replaceKey) {
      this.clearTimer(this.timer);
      this.timer = null;
      this.current = null;
      this.start(entry);
    } else if (entry.priority > this.current.priority) {
      this.clearTimer(this.timer);
      this.timer = null;
      this.current = null;
      this.start(entry);
    } else {
      this.pending.push(entry);
      this.pending.sort((a, b) => b.priority - a.priority || a.sequence - b.sequence);
    }
    return true;
  }

  start(entry) {
    this.current = entry;
    const { sequence: _sequence, ...message } = entry;
    this.deliver(message);
    this.timer = this.setTimer(() => {
      this.current = null;
      this.timer = null;
      const next = this.pending.shift();
      if (next) this.start(next);
    }, entry.durationMs);
  }

  clear() {
    if (this.timer) this.clearTimer(this.timer);
    this.timer = null;
    this.current = null;
    this.pending = [];
  }
}

module.exports = { SpeechQueue };

File: native-appkit/Sources/BlobfishNative/SpeechQueue.swift
import Foundation

struct SpeechMessage: Equatable {
    let text: String
    let event: String?
    let priority: Int
    let duration: TimeInterval
    let replaceKey: String?
    fileprivate let sequence: Int
}

final class SpeechQueue {
    private let deliver: (SpeechMessage) -> Void
    private let onIdle: () -> Void
    private(set) var current: SpeechMessage?
    private(set) var pending: [SpeechMessage] = []
    private var timer: Timer?
    private var sequence = 0

    init(deliver: @escaping (SpeechMessage) -> Void, onIdle: @escaping () -> Void) {
        self.deliver = deliver
        self.onIdle = onIdle
    }

    @discardableResult
    func enqueue(
        text: String,
        event: String?,
        priority: Int,
        duration: TimeInterval,
        replaceKey: String?
    ) -> Bool {
        guard !text.isEmpty else { return false }
        let entry = SpeechMessage(
            text: text,
            event: event,
            priority: priority,
            duration: duration,
            replaceKey: replaceKey,
            sequence: sequence
        )
        sequence += 1

        if let replaceKey {
            pending.removeAll { $0.replaceKey == replaceKey }
        }
        guard let current else {
            start(entry)
            return true
        }
        if let replaceKey, replaceKey == current.replaceKey {
            cancelCurrent()
            start(entry)
        } else if entry.priority > current.priority {
            cancelCurrent()
            start(entry)
        } else {
            pending.append(entry)
            pending.sort { left, right in
                left.priority == right.priority
                    ? left.sequence < right.sequence
                    : left.priority > right.priority
            }
        }
        return true
    }

    func clear() {
        timer?.invalidate()
        timer = nil
        current = nil
        pending.removeAll()
        onIdle()
    }

    func finishCurrent() {
        timer?.invalidate()
        timer = nil
        current = nil
        if pending.isEmpty {
            onIdle()
        } else {
            start(pending.removeFirst())
        }
    }

    private func cancelCurrent() {
        timer?.invalidate()
        timer = nil
        current = nil
    }

    private func start(_ entry: SpeechMessage) {
        current = entry
        deliver(entry)
        timer = Timer.scheduledTimer(withTimeInterval: entry.duration, repeats: false) { [weak self] _ in
            self?.finishCurrent()
        }
        if let timer { RunLoop.main.add(timer, forMode: .common) }
    }
}
